cogito/protocol.go

```diff
@@ -0,0 +1,184 @@
+	Env Environment
+}
+
+func NewCheckInput(in io.Reader) (CheckInput, error) {
+	var ci CheckInput
+
+	dec := json.NewDecoder(in)
+	dec.DisallowUnknownFields()
+	if err := dec.Decode(&ci); err != nil {
+		return ci, fmt.Errorf("parsing JSON from stdin: %s", err)
+	}
+
+	if err := ci.Source.Init(); err != nil {
+		return ci, err
+	}
+
+	ci.Env.Fill()
+
+	return ci, nil
+}
+
+// Source is the "source:" block in a pipeline "resources:" block for the Cogito resource.
+type Source struct {
+	//
+	// Mandatory
+	//
+	Owner       string `json:"owner"`
+	Repo        string `json:"repo"`
+	AccessToken string `json:"access_token"` // SENSITIVE
+	//
+	// Optional
+	//
+	GChatWebHook  string `json:"gchat_webhook"` // SENSITIVE
+	LogLevel      string `json:"log_level"`
+	LogUrl        string `json:"log_url"` // DEPRECATED
+	ContextPrefix string `json:"context_prefix"`
+}
+
+// String renders Source, redacting the sensitive fields.
+func (src Source) String() string {
+	var bld strings.Builder
+
+	fmt.Fprintln(&bld, "owner:         ", src.Owner)
+	fmt.Fprintln(&bld, "repo:          ", src.Repo)
+	fmt.Fprintln(&bld, "access_token:  ", redact(src.AccessToken))
+	fmt.Fprintln(&bld, "gchat_webhook: ", redact(src.GChatWebHook))
+	fmt.Fprintln(&bld, "log_level:     ", src.LogLevel)
+	fmt.Fprint(&bld, "context_prefix: ", src.ContextPrefix)
+
+	return bld.String()
+}
+
+// redact returns a redacted version of s. If s is empty, it returns the empty string.
+func redact(s string) string {
+	if s != "" {
+		s = "***REDACTED***"
+	}
+	return s
+}
+
+// Init validates and applies defaults for Source.
+func (src *Source) Init() error {
+	//
+	// Validate mandatory fields.
+	//
+	var mandatory []string
+	if src.Owner == "" {
+		mandatory = append(mandatory, "owner")
+	}
+	if src.Repo == "" {
+		mandatory = append(mandatory, "repo")
+	}
+	if src.AccessToken == "" {
+		mandatory = append(mandatory, "access_token")
+	}
+	if len(mandatory) > 0 {
+		return fmt.Errorf("source: missing keys: %s", strings.Join(mandatory, ", "))
+	}
+
+	//
+	// Validate optional fields.
+	//
+
+	// Normally we would leave this validation directly to the logging package, but since
+	// the log level names are part of the Cogito API and predate the removal of ofcourse,
+	// we need to handle the mapping and the error message, to avoid confusing the user.
+	logAdapter := map[string]string{
+		"debug":  "debug",
+		"info":   "info",
+		"warn":   "warn",
+		"error":  "error",
+		"silent": "off", // different
+	}
+	if src.LogLevel != "" {
+		if _, ok := logAdapter[src.LogLevel]; !ok {
+			return fmt.Errorf("source: invalid log_level: %s", src.LogLevel)
+		}
+		src.LogLevel = logAdapter[src.LogLevel]
+	}
+
+	//
+	// Apply defaults.
+	//
+	if src.LogLevel == "" {
+		src.LogLevel = "info"
+	}
+
+	return nil
+}
+
+// Version is a JSON object with string fields. Part of the Concourse resource protocol.
+type Version struct {
+	Ref string `json:"ref"`
+}
```

Comment:
This is generally wrong.

From https://concourse-ci.org/implementing-resource-types.html#resource-check:
> `version` is a JSON object with string fields, used to uniquely identify an instance of the resource.

So the only constraint is that the values must be strings. Nothing is said about the presence of specific keys. And it's easy to find example of versions without a `ref`:
```sh
$ fly -t developers resource-versions -r third-party-master/pushprox-amd64.s3 --json -c 1
[
  {
    "id": 2966343650,
    "version": {
      "path": "final/pushprox/master/pushprox-proxy-v0.1.0-41-g9778f0e"
    },
    "enabled": true
  }
]
```

Given you're allowed to return an empty list, I don't think it makes sense to bother with supporting version in any form for cogito.

Reply:
I think that part of the confusion comes from the comment. What is part of the protocol is a JSON object with string fields, the fields are indeed opaque for Concourse.

Here the key "ref" is what I added in the first version of Cogito. It is true that written like this is seems that "ref" is part of the protocol. I will clarify.

As in the previous comment, it might be that previous Concourse versions did want something, I cannot remember.

I prefer to proceed as follows:

- First keep this refactoring a real refactoring: external behavior should stay the same.
- Then, once the migration is complete, try to remove the dummy version. One of the things that leave me a bit unease is what happens to a Concourse that has in the database for Cogito already the `{"ref": "dummy"}` version ...

Thanks for the attention to the specs, greatly appreciated @odormond

Reply:
Will be investigated in PCI-2617